### PR TITLE
Improve scripts/upload-table-config for safety + efficiency (+updates other scripts to use new patterns)

### PR DESCRIPTION
**Pull request by 0xdevalias**
- Bump `ts-node` so we can make use of new ESM support + shebang for scripts
  - https://github.com/TypeStrong/ts-node/issues/1007
- DRY up common admin helper script patterns into `scripts/lib/helpers`
- Save local backups of data before we modify it just in case #BadThingsHappen
- Run our DB updates within a transaction to avoid potential damage from race conditions
- Ensure we only update the deep key we want to change when writing back to the DB

### CODE REVIEW

File: scripts/lib/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+  const jsonData = JSON.stringify(data, null, 2);
+
+  writeFileSync(backupFilename, jsonData, "utf8");
+
+  console.log(`Saved backup (type='${type}') to ${backupFilename}`);
+};
+
+export const initFirebaseAdminApp = (projectId: string) => {
+  admin.initializeApp({
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply scripts/lib/helpers.ts:24]: [OOS/Future] Those duplicate locations can all be updated in future to make use of this new helper that abstracts it. But for now, that is OOS.

**Review by thecadams — APPROVED** [review]
Great improvement, really like the way the transaction makes the success/failure case so clean!

**Comment by 0xdevalias** [comment]
> Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: /Users/devalias/dev/sparkle/sparkle/jest.config.js
> require() of ES modules is not supported.
> require() of /Users/devalias/dev/sparkle/sparkle/jest.config.js from /Users/devalias/dev/sparkle/sparkle/node_modules/jest-config/build/readConfigFileAndSetRootDir.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
> Instead rename jest.config.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/devalias/dev/sparkle/sparkle/package.json.

File: scripts/create-jazz-bar.ts [hunk 2]
```diff
@@ -1,34 +1,55 @@
...
+  projectId,
+  name,
+  iframeUrl,
+  logoImageUrl,
+  iconUrl,
+  _profileQuestions,
+  _codeofConductQuestions,
+] = process.argv.slice(2);
+
+if (
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply scripts/create-jazz-bar.ts:36]: [OOS] didn't want to spend the time refactoring all of these scripts. Progressive improvements > perfection

File: scripts/duplicate-venue.ts [hunk 3]
```diff
@@ -1,48 +1,50 @@
...
-import serviceAccount from "./prodAccountKey.json";
-import "firebase/firestore";
 
-function usage() {
-  console.log(`
-${process.argv[1]}: Get venue details. Prints venue name, type and other details.
+import { initFirebaseAdminApp } from "./lib/helpers";
 
-Usage: node ${process.argv[1]} PROJECT_ID VENUE_ID TARGET_VENUE_ID
+const usage = () => {
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply scripts/duplicate-venue.ts:6]: NA: this is the pattern we're using to show our usage.. not a whole lot we could do to make it better.

File: scripts/get-visitors.ts [hunk 4]
```diff
@@ -1,45 +1,47 @@
...
-import "firebase/firestore";
+import { initFirebaseAdminApp } from "./lib/helpers";
 
-const CODES_FILE_PATH = "codes.txt";
-const codes = fs
-  .readFileSync(CODES_FILE_PATH)
-  .toString()
-  .split("\n")
-  .map((line) => line.toLowerCase());
+const usage = () => {
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply scripts/get-visitors.ts:7]: NA: this is the pattern we're using to show our usage.. not a whole lot we could do to make it better.

File: scripts/create-jazz-bar.ts [hunk 5]
```diff
@@ -1,34 +1,55 @@
...
+
 import { JazzbarVenue } from "../src/types/JazzbarVenue";
 import { VenueTemplate } from "../src/types/VenueTemplate";
 
-function usage() {
-  console.log(`
-${process.argv[1]}: Create jazz bar
+import { initFirebaseAdminApp } from "./lib/helpers";
+
+const usage = () => {
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply scripts/create-jazz-bar.ts:10]: NA: this is the pattern we're using to show our usage.. not a whole lot we could do to make it better.

File: scripts/upload-table-config.ts [hunk 6]
```diff
@@ -1,7 +1,38 @@
...
 import admin from "firebase-admin";
-import serviceAccount from "./prodAccountKey.json";
-import "firebase/firestore";
+
 import { Table } from "../src/types/Table";
+import { Venue } from "../src/types/Venue";
+
+import { initFirebaseAdminApp, makeSaveToBackupFile } from "./lib/helpers";
+
+const usage = () => {
```
[the remark this answers is not in the document]
Comment (0xdevalias) [reply scripts/upload-table-config.ts:10]: NA: this is the pattern we're using to show our usage.. not a whole lot we could do to make it better.

**Review by thecadams — APPROVED** [review]
Nothing concerning in here and a lot of nice stuff, like the usage message checks, backups... LGTM.

**Pull request merged**